File: automcs150-frontend/middleware.ts
import { NextResponse } from 'next/server';
import { jwtVerify } from 'jose';
import type { NextRequest } from 'next/server';

const JWT_SECRET_KEY = process.env.JWT_SECRET_KEY!;
const secret = new TextEncoder().encode(JWT_SECRET_KEY);

export async function middleware(request: NextRequest) {

  const { pathname } = request.nextUrl;
  const token = request.cookies.get('accesstoken')?.value;

  if (!token) {
    return NextResponse.redirect(new URL('/login', request.url));
  }
  try {
    // 👇 Verify and decode token
    const { payload } = await jwtVerify(token, secret);

    // Access email from the payload
    const { email, firstName, lastName, isAdmin } = payload;

    if (!isAdmin && pathname.includes("admin")) {
      return NextResponse.redirect(new URL('/dashboard', request.url));
    }
    // Token is valid, continue
    return NextResponse.next();
  } catch (err: any) {
    // 👇 If token is expired or invalid
    console.error('JWT error:', err.message);
    return NextResponse.redirect(new URL('/login', request.url));
  }
}

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - _next/static (static files)
     * - favicon.ico (favicon file)
     */
    '/((?!login|register|_next/static|api/auth|favicon.ico|.*\\.jpg$).*)'
  ]
}

File: automcs150-backend/middleware.py
# middleware.py
from flask import request, jsonify
import jwt
import flask

EXEMPT_ROUTES = ["/api/auth/login", "/api/auth/register"]


def middleware(app):
    @app.before_request
    def check_jwt_token():
        if request.path in EXEMPT_ROUTES:
            return

        token = request.headers.get("Authorization")
        print(request.headers)
        if token is None or not token.startswith("Bearer "):
            return jsonify({"error": "Missing or invalid token"}), 401

        try:
            jwt_token = token.split(" ")[1]
            decoded = jwt.decode(
                jwt_token, app.config["JWT_SECRET_KEY"], algorithms=["HS256"]
            )
            request.user = decoded
        except jwt.ExpiredSignatureError:
            return jsonify({"error": "Token has expired"}), 401
        except jwt.InvalidTokenError:
            return jsonify({"error": "Invalid token"}), 401
